### PR TITLE
Feature/policy parse

### PR DESCRIPTION
**Pull request by bplatz**
This addresses issue #348 and updates the early policy parsing code to make it more flexible. The initial parsing code was built to demonstrate view permissions quickly, not made more generically to address other permissions (specifically modify for now, others later) which will be addressed in #296 with this foundation.

It also sets up using `targetNodes` for enforcement (not yet done) - previously this was only used to target `allNodes` as an ideal way to establish a 'root' access policy.

Policy parsing, be it for classes or nodes, now returns two-tuples as it reaches into deeper logic that represent:
[key-sequence-in-final-policy-map policy-function-for-allow-flake?]

For example, view permission for class :ex/User (id: 111) and predicate :ex/ssn (id 222) would end up in a policy map like:
`{:f/view {:class {111 {222 policy-function-for-allow-flake?]}}}}`

and therefore in parsing would return the two-tuple like so, such that the key seq can be used with `assoc-in`:
[`[:f/view :class 111 222]` policy-function-for-allow-flake?]

The other change is that instead of only holding only a bytecode function in policy-function-for-allow-flake?, it now holds a map that contains the previous function in the `:function` key. This is not needed by `allow-flake?` but can help anyone debugging what the bytecode function is doing by including the original policy metadata for that function - this comes as the cost of more memory, but should be minimal.

### CODE REVIEW

File: test/fluree/db/policy/parsing.clj [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          sid-userRole @(fluree/promise-wrap (dbproto/-subid db :ex/userRole))
+          sid-alice    @(fluree/promise-wrap (dbproto/-subid db alice-did))
+          sid-User     @(fluree/promise-wrap (dbproto/-subid db :ex/User))
+          sid-ssn      @(fluree/promise-wrap (dbproto/-subid db :schema/ssn))
+          ;; create optimized policy map for ex:userRole
+          policy-alice @(fluree/promise-wrap (policy/policy-map db alice-did :ex/userRole nil))
+          policy-root  @(fluree/promise-wrap (policy/policy-map db root-did :ex/rootRole nil))]
+
+      ;; look at  policy for user
+      (is (= (replace-policy-fns policy-alice)              ;; replace compiled functions for data comparisons below
```
Comment (dpetran) [inline test/fluree/db/policy/parsing.clj:97]: It looks like you've swapped the `expected` and `actual` positions here, just fyi that the test runner produces more readable failures if you reverse the arguments to `=`.

**Comment by dpetran** [comment]
This replaces the code in `fluree.db.permissions` and `fluree.db.premissions-validate`, or does it supplement it? Looking at it I think it's a replacement, but I want to confirm.

**Review by dpetran — APPROVED** [review]
:+1:

**Comment by bplatz** [comment]
> This replaces the code in `fluree.db.permissions` 

yes

> and `fluree.db.premissions-validate`

no

Good question, as it made me realize we now have policy-validate and permissions-validate. I'd like to see all of this live under `fluree.db.policy.*` and considered making that change but thought it would the PR harder to see what changed. Along with that, I think updating the naming of the two `*-validate` (or even consolidating them) would be good to do.

File: test/fluree/db/policy/parsing.clj (hunk 1 above)
Comment (bplatz) [reply test/fluree/db/policy/parsing.clj]: Fixed, thanks!

**Review by zonotope — APPROVED** [review]
This was already approved while I was reviewing. It looks good to me, but I have some very minor suggestions and a question, all of which can be addressed later.

**Comment by chrisdostert** [comment]
> It also sets up using targetNodes for enforcement (not yet done)

@bplatz sorry to ask here but do you know if `targetNode` enforcement is working in v3? I am trying to get it working per [the docs](https://next.developers.flur.ee/docs/learn/foundations/data-access-control/#ftargetnode) but am having no luck. Trying to bottom out why has led me here.

**Pull request merged**